# Segmenting and Clustering Neighborhoods in the city of Toronto, Canada

## Assignment Part 1: Create dataframe to match

### Import all the required libraries

Pandas, urlopen and BeautifulSoup to be used

In [71]:
# import pandas library for data analsysis
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# import library to handle requests
import requests

# tranform JSON file into a pandas dataframe
from pandas import json_normalize

# import numpy
import numpy as np

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import urlopen to access internet data
from urllib.request import urlopen

# import Beautiful soup to parse html page
from bs4 import BeautifulSoup

#import folium for generating map
import folium

# import k-means from clustering stage
from sklearn.cluster import KMeans

# confirm that all libraries have been imported
print('Libraries imported!')

Libraries imported!


### Obtain data from specified url

Will use urlopen and read() to get the html codes of the link. The html codes will then be parsed through BeautifulSoup to allow me to access the specific table.

In [72]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
# use urlopen to open url and read() to obtain html codes
html = urlopen(url).read()
# use BeautifulSoup to parse html
raw_data = BeautifulSoup(html,'html.parser')

### Obtain column header from web table using the html tags 'th'

The table column names are found to be wrapped within the html tags 'th'. So, I'll go ahead and get it and feed it into a column_header list.

In [98]:
column_tags = raw_data('th')
# create empty list for column header
column_header = []
# iterate across tags, get the text, strip \n at the end and append to column header list
for tags in column_tags:
    header = tags.text.rstrip()
    column_header.append(header)

### Initialize dataframe using the column names

The dataframe is then initialised using the column_header list, dropping 'Canadian postal codes' as it is not required as part of the assignment.

In [99]:
df = pd.DataFrame(columns = column_header).drop(columns='Canadian postal codes')
df

Empty DataFrame
Columns: [Postal Code, Borough, Neighborhood]
Index: []

### Obtain row data from each row using the html tags 'tr'

Each row of information is found between the html tags 'tr'

In [100]:
row_tags = raw_data('tr')
# iterate across each row tags except 0 as this is the column header row and the last 4 rows which is not relevant
for tags in row_tags[1:len(row_tags)-4]:
        row_postal_code = tags('td')[0].text.rstrip()
        row_borough = tags('td')[1].text.rstrip()
        row_neighborhood = tags('td')[2].text.rstrip()
        #ignore row where borough is not assigned
        if row_borough != 'Not assigned':
            #append each borough assigned row into dataframe
            df = df.append({'Postal Code':row_postal_code,
                            'Borough':row_borough,
                            'Neighborhood':row_neighborhood},ignore_index=True)
df.head(15)

Postal Code           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4         Queen's Park, Ontario Provincial Government  
5                                    Islington Avenue  
6                                      Malvern, Rouge  
7                                           Don Mills  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...  
12             Rouge Hill, Port Union, Highland Creek  
13                                          Don Mills  
14                                   Woodbine Heights

### Explore how many boroughs and postal code within each borough

In [101]:
print('There are {} boroughs, {} neighborhood and {} postal codes in Toronto.'.
      format(len(df['Borough'].unique()),len(df['Neighborhood'].unique()),len(df['Postal Code'].unique())))

There are 10 boroughs, 98 neighborhood and 103 postal codes in Toronto.


Seems like there are duplicates of neighborhood in a couple of postal codes in Toronto. Let's find out what they are

In [102]:
df.groupby(['Borough','Neighborhood']).count().sort_values(by=['Postal Code'],ascending=False).head()

Postal Code
Borough         Neighborhood                                         
North York      Downsview                                           4
                Don Mills                                           2
                Willowdale                                          2
Central Toronto Davisville                                          1
North York      North Park, Maple Leaf Park, Upwood Park            1

Seems like Donsview, Don Mills and Willowdale has multiple postal codes. To make it distinct we can append the postal codes at the end of the neighborhoods

In [103]:
df['Neighborhood'] = df[['Neighborhood','Postal Code']].apply(lambda x: '-'.join(x),axis=1)

Let's check again

In [104]:
df.groupby(['Borough','Neighborhood']).count().sort_values(by=['Postal Code'],ascending=False).head()

Postal Code
Borough         Neighborhood                                     
Central Toronto Davisville North-M4P                            1
North York      Lawrence Manor, Lawrence Heights-M6A            1
Scarborough     Agincourt-M1S                                   1
North York      York Mills, Silver Hills-M2L                    1
                York Mills West-M2P                             1

Now I have distinct neighborhoods, it will allow me to map out venues in distinctive postal code areas. Another solution is to remove rows of duplicate neighborhoods but I decided to keep it going forwards. 

In [105]:
print(df.shape)

(103, 3)


## Assignment Part II: Combining previous dataframe with geospatial data

I am going to use the geosptial_data.csv instead of Geocoder due to inability to obtain data from the library

### Download csv file and read geospatial data

In [11]:
!wget -q -O Geospatial_data.csv https://cocl.us/Geospatial_data    

### Read geospatial.csv file

In [106]:
geo_spatial = pd.read_csv('Geospatial_data.csv')
geo_spatial.head(10)

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476
5         M1J  43.744734 -79.239476
6         M1K  43.727929 -79.262029
7         M1L  43.711112 -79.284577
8         M1M  43.716316 -79.239476
9         M1N  43.692657 -79.264848

### Merge geo_spatial data with main df

In [107]:
df = pd.merge(df,geo_spatial,how='left',on=['Postal Code'])

In [108]:
df2 = df
df2['Neighborhood'] = df2['Neighborhood'].str[:-4]
print(df.shape)
df2.head(15)

(103, 5)


Postal Code           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   

                                         Neighborhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                           Regent Park, Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
5                                    Islington Avenue  43.667856 -79.532242  
6                                      Malvern, Rouge  43.806686 -79.194353  
7                                           Don Mills  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724  
12             Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
13                                          Don Mills  43.725900 -79.340923  
14                                   Woodbine Heights  43.695344 -79.318389

## Assignment Part III: Exploration and clustering of neighborhoods in Toronto

### Use geopy to obtain longitude and latitude of Toronto city for use in folium

In [84]:
from geopy.geocoders import Nominatim

Define instance for geocoder and name agent at 'Toronto_explorer'

In [85]:
address = 'Toronto City, Canada'

geolocator = Nominatim(user_agent="Toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


Create map of Toronto City with neighborhoods superimposed on top. The neighborhoods corresponds to each postal code and note that neighborhoods Donsview, Don Mills and Willowdale has multiple postal codes.

In [86]:
# create map of Toronto City using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

I am going to explore all the neighborhoods and segment them by using the Foursquare API

### Explore one of Toronto's neighborhood

Let's explore one of the neighbourhood(s) in Toronto. We will use the first one. Let's get the latitude and longitude for that.

In [20]:
neighborhood_latitude = df.loc[0,'Latitude']
neighborhood_longitude = df.loc[0,'Longitude']
neighborhood_name = df.loc[0,'Neighborhood']

print('The {} neighborhood(s) has the latitude of {} and longitude of {}'.format(neighborhood_name,
                                                                                neighborhood_latitude,
                                                                                neighborhood_longitude))

The Parkwoods-M3A neighborhood(s) has the latitude of 43.7532586 and longitude of -79.3296565


Let's get the top 100 venues in that area with a 1000 metre radius. Create a GET url and obtain json

In [52]:
radius = 1000
LIMIT = 100
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, neighborhood_latitude, neighborhood_longitude, VERSION, radius, LIMIT)
results = requests.get(url).json()

Borrowing the get_category_type function from earlier lab.

In [24]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Clean and restructure json data into a dataframme

In [25]:
venues = results['response']['groups'][0]['items']

nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues = nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name            categories        lat        lng
0    Allwyn's Bakery  Caribbean Restaurant  43.759840 -79.324719
1    Brookbanks Park                  Park  43.751976 -79.332140
2        Tim Hortons                  Café  43.760668 -79.326368
3                A&W  Fast Food Restaurant  43.760643 -79.326865
4  Bruno's valu-mart         Grocery Store  43.746143 -79.324630

Now I have the venues for one neighborhood. How many venues were returned by Foursquare?

In [26]:
print('{} venues were returned by Foursquare for the {} neighborhood.'.format(nearby_venues.shape[0],
                                                                             neighborhood_name))

29 venues were returned by Foursquare for the Parkwoods-M3A neighborhood.


### Explore all of Toronto's neighborhood

Borrowing function from earlier labs to repeat the above steps for all neighborhoods in Toronto

In [27]:
def getNearbyVenues(names, latitudes, longitudes, radius=1000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()['response']['groups'][0]['items']
                   
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Use above defined function and run it for all neighborhoods in Toronto and put it to *toronto_venues*

In [28]:
print(df.shape)
toronto_venues = getNearbyVenues(names=df['Neighborhood'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                  )


(103, 5)
Parkwoods-M3A
Victoria Village-M4A
Regent Park, Harbourfront-M5A
Lawrence Manor, Lawrence Heights-M6A
Queen's Park, Ontario Provincial Government-M7A
Islington Avenue-M9A
Malvern, Rouge-M1B
Don Mills-M3B
Parkview Hill, Woodbine Gardens-M4B
Garden District, Ryerson-M5B
Glencairn-M6B
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale-M9B
Rouge Hill, Port Union, Highland Creek-M1C
Don Mills-M3C
Woodbine Heights-M4C
St. James Town-M5C
Humewood-Cedarvale-M6C
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood-M9C
Guildwood, Morningside, West Hill-M1E
The Beaches-M4E
Berczy Park-M5E
Caledonia-Fairbanks-M6E
Woburn-M1G
Leaside-M4G
Central Bay Street-M5G
Christie-M6G
Cedarbrae-M1H
Hillcrest Village-M2H
Bathurst Manor, Wilson Heights, Downsview North-M3H
Thorncliffe Park-M4H
Richmond, Adelaide, King-M5H
Dufferin, Dovercourt Village-M6H
Scarborough Village-M1J
Fairview, Henry Farm, Oriole-M2J
Northwood Park, York University-M3J
East Toronto-M4J
Harbourfront 

Checking the size of the resulting dataframe

In [29]:
print(toronto_venues.shape)
toronto_venues.head()

(4911, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Parkwoods-M3A              43.753259              -79.329656   
1  Parkwoods-M3A              43.753259              -79.329656   
2  Parkwoods-M3A              43.753259              -79.329656   
3  Parkwoods-M3A              43.753259              -79.329656   
4  Parkwoods-M3A              43.753259              -79.329656   

               Venue  Venue Latitude  Venue Longitude        Venue Category  
0    Allwyn's Bakery       43.759840       -79.324719  Caribbean Restaurant  
1    Brookbanks Park       43.751976       -79.332140                  Park  
2        Tim Hortons       43.760668       -79.326368                  Café  
3                A&W       43.760643       -79.326865  Fast Food Restaurant  
4  Bruno's valu-mart       43.746143       -79.324630         Grocery Store

Group the venues by neighborhood to check how many venues are found per neighborhood

In [30]:
print(toronto_venues.groupby('Neighborhood').count().shape)
toronto_venues.groupby('Neighborhood').count()

(102, 6)


Neighborhood Latitude  \
Neighborhood                                                                
Agincourt-M1S                                                          44   
Alderwood, Long Branch-M8W                                             27   
Bathurst Manor, Wilson Heights, Downsview North...                     30   
Bayview Village-M2K                                                    14   
Bedford Park, Lawrence Manor East-M5M                                  41   
Berczy Park-M5E                                                       100   
Birch Cliff, Cliffside West-M1N                                        14   
Brockton, Parkdale Village, Exhibition Place-M6K                      100   
Business reply mail Processing Centre-M7Y                              49   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Caledonia-Fairbanks-M6E                                                24   
Canada Post Gateway Processing Centre-M7R                              55   
Cedarbrae-M1H                                                          30   
Central Bay Street-M5G                                                100   
Christie-M6G                                                          100   
Church and Wellesley-M4Y                                              100   
Clarks Corners, Tam O'Shanter, Sullivan-M1T                            33   
Cliffside, Cliffcrest, Scarborough Village West...                     12   
Commerce Court, Victoria Hotel-M5L                                    100   
Davisville North-M4P                                                  100   
Davisville-M4S                                                        100   
Del Ray, Mount Dennis, Keelsdale and Silverthor...                     17   
Don Mills-M3B                                                          32   
Don Mills-M3C                                                          41   
Dorset Park, Wexford Heights, Scarborough Town ...                     46   
Downsview-M3K                                                          21   
Downsview-M3L                                                           9   
Downsview-M3M                                                           4   
Downsview-M3N                                                          35   
Dufferin, Dovercourt Village-M6H                                       67   
East Toronto-M4J                                                       95   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                     18   
Fairview, Henry Farm, Oriole-M2J                                       44   
First Canadian Place, Underground city-M5X                            100   
Forest Hill North & West-M5P                                           46   
Garden District, Ryerson-M5B                                          100   
Glencairn-M6B                                                          34   
Golden Mile, Clairlea, Oakridge-M1L                                    27   
Guildwood, Morningside, West Hill-M1E                                  25   
Harbourfront East, Union Station, Toronto Islan...                    100   
High Park, The Junction South-M6P                                     100   
Hillcrest Village-M2H                                                  20   
Humber Summit-M9L                                                      10   
Humberlea, Emery-M9M                                                    8   
Humewood-Cedarvale-M6C                                                 26   
India Bazaar, The Beaches West-M4L                                     82   
Islington Avenue-M9A                                                   12   
Kennedy Park, Ionview, East Birchmount Park-M1K                        30   
Kensington Market, Chinatown, Grange Park-M5T                         100   
Kingsview Village, St. Phillips, Martin Grove G...                     16   
Lawrence Manor, Lawrence Heights-M6A            

Seems like there is a neighborhood which has not return venues at all. Which one is it?

In [31]:
venues_grouped = toronto_venues.groupby('Neighborhood').count().reset_index()
missing_neigh=df[~df.Neighborhood.isin(venues_grouped.Neighborhood)]
missing_neigh

Postal Code      Borough     Neighborhood   Latitude  Longitude
95         M1X  Scarborough  Upper Rouge-M1X  43.836125 -79.205636

Let's confirm by trying a Foursquare API call

In [32]:
missing_latitude = missing_neigh.loc[95,'Latitude']
missing_longitude = missing_neigh.loc[95,'Longitude']
radius = 1000
LIMIT = 100
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, missing_latitude, missing_longitude, VERSION, radius, LIMIT)
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ec19e2b9388d7001c2709d8'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Rouge',
  'headerFullLocation': 'Rouge, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 0,
  'suggestedBounds': {'ne': {'lat': 43.84512470900002,
    'lng': -79.19318232223782},
   'sw': {'lat': 43.827124690999995, 'lng': -79.21808987776217}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': []}]}}

There was indeed no results. We'll have to ensure that this row of Postal Code M1X and Neighborhood Upper Rouge is not reflected later.

Investigate how many unique categories there are from all returned values

In [33]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 327 uniques categories.


Analyzing each neighborhood through one hot encoding

In [63]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = list(toronto_onehot)
fixed_columns.insert(0,fixed_columns.pop(fixed_columns.index('Neighborhood')))
toronto_onehot = toronto_onehot[fixed_columns]

# Clean up neighborhood column as venues data has been retrieved
toronto_onehot['Neighborhood'] = toronto_onehot['Neighborhood'].str[:-4]
toronto_onehot.head()

Neighborhood  Accessories Store  Afghan Restaurant  Airport  Airport Lounge  \
0    Parkwoods                  0                  0        0               0   
1    Parkwoods                  0                  0        0               0   
2    Parkwoods                  0                  0        0               0   
3    Parkwoods                  0                  0        0               0   
4    Parkwoods                  0                  0        0               0   

   American Restaurant  Amphitheater  Animal Shelter  Antique Shop  Aquarium  \
0                    0             0               0             0         0   
1                    0             0               0             0         0   
2                    0             0               0             0         0   
3                    0             0               0             0         0   
4                    0             0               0             0         0   

   Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0            0           0                    0                 0   
1            0           0                    0                 0   
2            0           0                    0                 0   
3            0           0                    0                 0   
4            0           0                    0                 0   

   Athletics & Sports  Auto Dealership  Auto Garage  Auto Workshop  \
0                   0                0            0              0   
1                   0                0            0              0   
2                   0                0            0              0   
3                   0                0            0              0   
4                   0                0            0              0   

   Automotive Shop  BBQ Joint  Baby Store  Badminton Court  Bagel Shop  \
0                0          0           0                0           0   
1                0          0           0                0           0   
2                0          0           0                0           0   
3                0          0           0                0           0   
4                0          0           0                0           0   

   Bakery  Bank  Bar  Baseball Field  Baseball Stadium  Basketball Stadium  \
0       0     0    0               0                 0                   0   
1       0     0    0               0                 0                   0   
2       0     0    0               0                 0                   0   
3       0     0    0               0                 0                   0   
4       0     0    0               0                 0                   0   

   Beach  Beach Bar  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0      0          0         0           0                   0          0   
1      0          0         0           0                   0          0   
2      0          0         0           0                   0          0   
3      0          0         0           0                   0          0   
4      0          0         0           0                   0          0   

   Bistro  Board Shop  Bookstore  Boutique  Bowling Alley  \
0       0           0          0         0              0   
1       0           0          0         0              0   
2       0           0          0         0              0   
3       0           0          0         0              0   
4       0           0          0         0              0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  Bridge  \
0                     0               0        0            0       0   
1                     0               0        0            0       0   
2                     0               0        0            0       0   
3                     0               0        0            0       0   
4                     0               0        0            0       0   

   Bubble Tea Shop  B

What is the shape of the new toronto dataframe?

In [64]:
toronto_onehot.shape

(4911, 327)

The shape matches the total number of venue categories and number of venues found. Now let's group them by the neighborhood to get an idea of the frequency of venue occurences in each category

In [65]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Afghan Restaurant  Airport  Airport Lounge  American Restaurant  \
0                0.0      0.0             0.0             0.022727   
1                0.0      0.0             0.0             0.000000   
2                0.0      0.0             0.0             0.000000   
3                0.0      0.0             0.0             0.000000   
4                0.0      0.0             0.0             0.024390   

   Amphitheater  Animal Shelter  Antique Shop  Aquarium  Art Gallery  \
0           0.0             0.0           0.0       0.0          0.0   
1           0.0             0.0           0.0       0.0          0.0   
2           0.0             0.0           0.0       0.0          0.0   
3           0.0             0.0           0.0       0.0          0.0   
4           0.0             0.0           0.0       0.0          0.0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0         0.0                  0.0          0.022727                 0.0   
1         0.0                  0.0          0.000000                 0.0   
2         0.0                  0.0          0.000000                 0.0   
3         0.0                  0.0          0.000000                 0.0   
4         0.0                  0.0          0.000000                 0.0   

   Auto Dealership  Auto Garage  Auto Workshop  Automotive Shop  BBQ Joint  \
0              0.0          0.0            0.0              0.0        0.0   
1              0.0          0.0            0.0              0.0        0.0   
2              0.0          0.0            0.0              0.0        0.0   
3              0.0          0.0            0.0              0.0        0.0   
4              0.0          0.0            0.0              0.0        0.0   

   Baby Store  Badminton Court  Bagel Shop    Bakery      Bank  Bar  \
0     0.00000         0.022727     0.00000  0.045455  0.022727  0.0   
1     0.00000         0.000000     0.00000  0.000000  0.000000  0.0   
2     0.00000         0.000000     0.00000  0.000000  0.066667  0.0   
3     0.00000         0.000000     0.00000  0.000000  0.142857  0.0   
4     0.02439         0.000000     0.02439  0.024390  0.048780  0.0   

   Baseball Field  Baseball Stadium  Basketball Stadium  Beach  Beach Bar  \
0             0.0               0.0                 0.0    0.0        0.0   
1             0.0               0.0                 0.0    0.0        0.0   
2             0.0               0.0                 0.0    0.0        0.0   
3             0.0               0.0                 0.0    0.0        0.0   
4             0.0               0.0                 0.0    0.0        0.0   

   Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  Bistro  Board Shop  \
0       0.0         0.0                 0.0        0.0     0.0         0.0   
1       0.0         0.0                 0.0        0.0     0.0         0.0   
2       0.0         0.0                 0.0        0.0     0.0         0.0   
3       0.0         0.0                 0.0        0.0     0.0         0.0   
4       0.0         0.0                 0.0        0.0     0.0         0.0   

   Bookstore  Boutique  Bowling Alley  Brazilian Restaurant  Breakfast Spot  \
0        0.0       0.0            0.0                   0.0        0.022727   
1        0.0       0.0            0.0                   0.0        0.000000   
2        0.0       0.0            0.0                   0.0        0.000000   
3        0.0       0.0            0.0                   0.0        0.000000   
4        0.0       0.0            0.0                   0.0      

Let's check the shape to make sure it is valid

In [66]:
toronto_grouped.shape

(97, 327)

The shape corresponds to the main df shape of 103 minus the neighborhood Upper Rouge which did not return any result from Foursquare APO

Borrowing function to rate the top 10 location for each neighborhood

In [67]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now to create a dataframe of neighborhoods with the top 10 most common venue categories

In [87]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
# match new dataframe 'Neighborhood' column with the toronto_grouped dataframe
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

print(neighborhoods_venues_sorted.shape)
neighborhoods_venues_sorted.head()

(97, 11)


Neighborhood 1st Most Common Venue  \
0                                        Agincourt    Chinese Restaurant   
1                           Alderwood, Long Branch        Discount Store   
2  Bathurst Manor, Wilson Heights, Downsview North           Pizza Place   
3                                  Bayview Village   Japanese Restaurant   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0         Shopping Mall                Bakery  Caribbean Restaurant   
1              Pharmacy                  Park     Convenience Store   
2                  Park           Coffee Shop                  Bank   
3         Grocery Store           Gas Station                  Bank   
4           Coffee Shop                  Bank                  Park   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0       Motorcycle Shop                  Bank           Coffee Shop   
1           Pizza Place           Coffee Shop                   Pub   
2              Pharmacy      Community Center   Fried Chicken Joint   
3    Chinese Restaurant            Restaurant         Shopping Mall   
4        Sandwich Place  Fast Food Restaurant            Bagel Shop   

  8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
0      Sushi Restaurant           Supermarket  Latin American Restaurant  
1        Sandwich Place                 Trail              Shopping Mall  
2      Sushi Restaurant           Supermarket   Mediterranean Restaurant  
3                  Café                  Park                      Trail  
4                Bakery             Juice Bar           Sushi Restaurant

### Now time to cluster these neighborhoods

Run k-means to cluster the neighborhood into 6 clusters.

In [88]:
# set number of clusters
kclusters = 6

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([5, 3, 5, 5, 2, 2, 3, 2, 2, 2])

Now lets combine the cluster labels with the top 10 venues

In [89]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

# create a merge dataframe but drop
toronto_merged = df2.drop(index=95)

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656               3                  Park   
1  43.725882 -79.315572               2           Coffee Shop   
2  43.654260 -79.360636               2           Coffee Shop   
3  43.718518 -79.464763               2        Clothing Store   
4  43.662301 -79.389494               2           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0         Shopping Mall     Convenience Store               Pharmacy   
1  Gym / Fitness Center          Hockey Arena  Portuguese Restaurant   
2               Theater                  Café             Restaurant   
3  Fast Food Restaurant           Coffee Shop             Restaurant   
4      Sushi Restaurant   Japanese Restaurant                   Park   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0               Bus Stop     Fish & Chips Shop           Supermarket   
1                 Lounge                  Park            Playground   
2                   Park                 Diner    Italian Restaurant   
3  Vietnamese Restaurant          Dessert Shop   Fried Chicken Joint   
4        Thai Restaurant             Gastropub      Ramen Restaurant   

  8th Most Common Venue   9th Most Common Venue 10th Most Common Venue  
0     Food & Drink Shop    Fast Food Restaurant           Skating Rink  
1           Pizza Place             Golf Course            Men's Store  
2                   Pub                  Bakery         Breakfast Spot  
3      Sushi Restaurant  Furniture / Home Store            Event Space  
4                  Café      Italian Restaurant         Ice Cream Shop

Now let's visualise the clusters

In [90]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examining the clusters

__Cluster 1__

In [91]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1,2] + list(range(6, toronto_merged.shape[1]))]]

Borough                                       Neighborhood  \
8     East York                    Parkview Hill, Woodbine Gardens   
18  Scarborough                  Guildwood, Morningside, West Hill   
34   North York                    Northwood Park, York University   
38  Scarborough        Kennedy Park, Ionview, East Birchmount Park   
51  Scarborough    Cliffside, Cliffcrest, Scarborough Village West   
64         York                                             Weston   
70    Etobicoke                                          Westmount   
77    Etobicoke  Kingsview Village, St. Phillips, Martin Grove ...   
89    Etobicoke  South Steeles, Silverstone, Humbergate, Jamest...   

   1st Most Common Venue 2nd Most Common Venue   3rd Most Common Venue  \
8                Brewery           Pizza Place    Fast Food Restaurant   
18           Pizza Place                  Bank    Fast Food Restaurant   
34           Coffee Shop           Pizza Place  Furniture / Home Store   
38        Discount Store    Chinese Restaurant             Coffee Shop   
51           Pizza Place        Ice Cream Shop                   Beach   
64           Pizza Place         Train Station             Coffee Shop   
70           Pizza Place           Gas Station          Breakfast Spot   
77              Pharmacy                  Bank            Intersection   
89         Grocery Store           Pizza Place                Pharmacy   

   4th Most Common Venue       5th Most Common Venue 6th Most Common Venue  \
8   Gym / Fitness Center  Construction & Landscaping    Athletics & Sports   
18           Coffee Shop            Greek Restaurant              Bus Line   
34            Restaurant                  Sports Bar    Miscellaneous Shop   
38           Pizza Place        Fast Food Restaurant         Grocery Store   
51            Sports Bar                Burger Joint                  Park   
64   Fried Chicken Joint                    Pharmacy          Skating Rink   
70          Intersection   Middle Eastern Restaurant           Supermarket   
77              Bus Line                 Supermarket       Supplement Shop   
89              Gym Pool        Caribbean Restaurant                  Park   

        7th Most Common Venue 8th Most Common Venue   9th Most Common Venue  \
8              Soccer Stadium        Breakfast Spot                    Café   
18                 Sports Bar        Breakfast Spot           Smoothie Shop   
34  Middle Eastern Restaurant         Metro Station    Fast Food Restaurant   
38             Sandwich Place          Intersection      Light Rail Station   
51  Cajun / Creole Restaurant        Hardware Store          Farmers Market   
64             Sandwich Place                 Diner  Furniture / Home Store   
70         Chinese Restaurant           Golf Course      Golf Driving Range   
77              Shopping Mall            Beer Store             Gas Station   
89             Sandwich Place            Beer Store     Fried Chicken Joint   

   10th Most Common Venue  
8               Gastropub  
18            Supermarket  
34            Frame Store  
38                   Bank  
51      Electronics Store  
64          Grocery Store  
70         Sandwich Place  
77     Chinese Restaurant  
89           Liquor Store

Cluster 1 seems to cover areas which generally has pizza place and fast food restautant. Looking at the map, it seems to be mostly outside Toronto City centre hence the assumption is that cluster 1 is mostly __Food and Beverage__ areas.

__Cluster 2__

In [92]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1,2] + list(range(6, toronto_merged.shape[1]))]]

Borough                                       Neighborhood  \
12   Scarborough             Rouge Hill, Port Union, Highland Creek   
101    Etobicoke  Old Mill South, King's Mill Park, Sunnylea, Hu...   

    1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
12         Breakfast Spot          Burger Joint                  Park   
101                  Park    Italian Restaurant         Shopping Mall   

    4th Most Common Venue        5th Most Common Venue 6th Most Common Venue  \
12             Playground           Italian Restaurant           Fish Market   
101        Ice Cream Shop  Eastern European Restaurant  Gym / Fitness Center   

    7th Most Common Venue        8th Most Common Venue 9th Most Common Venue  \
12                   Farm  Eastern European Restaurant           Flower Shop   
101           Dry Cleaner          Dumpling Restaurant     Electronics Store   

    10th Most Common Venue  
12       Electronics Store  
101   Ethiopian Restaurant

Cluster 2 covers smaller area of park and smaller food joints. Cluster 2 can be assumed to be __Special Community__ areas.

__Cluster 3__

In [93]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1,2] + list(range(6, toronto_merged.shape[1]))]]

Borough                                       Neighborhood  \
1          North York                                   Victoria Village   
2    Downtown Toronto                          Regent Park, Harbourfront   
3          North York                   Lawrence Manor, Lawrence Heights   
4    Downtown Toronto        Queen's Park, Ontario Provincial Government   
7          North York                                          Don Mills   
9    Downtown Toronto                           Garden District, Ryerson   
13         North York                                          Don Mills   
15   Downtown Toronto                                     St. James Town   
16               York                                 Humewood-Cedarvale   
19       East Toronto                                        The Beaches   
20   Downtown Toronto                                        Berczy Park   
24   Downtown Toronto                                 Central Bay Street   
25   Downtown Toronto                                           Christie   
30   Downtown Toronto                           Richmond, Adelaide, King   
31       West Toronto                       Dufferin, Dovercourt Village   
33         North York                       Fairview, Henry Farm, Oriole   
35          East York                                       East Toronto   
36   Downtown Toronto  Harbourfront East, Union Station, Toronto Islands   
37       West Toronto                           Little Portugal, Trinity   
41       East Toronto                       The Danforth West, Riverdale   
42   Downtown Toronto           Toronto Dominion Centre, Design Exchange   
43       West Toronto       Brockton, Parkdale Village, Exhibition Place   
47       East Toronto                     India Bazaar, The Beaches West   
48   Downtown Toronto                     Commerce Court, Victoria Hotel   
52         North York                            Willowdale, Newtonbrook   
54       East Toronto                                    Studio District   
55         North York                  Bedford Park, Lawrence Manor East   
59         North York                                         Willowdale   
61    Central Toronto                                      Lawrence Park   
62    Central Toronto                                           Roselawn   
67    Central Toronto                                   Davisville North   
68    Central Toronto                           Forest Hill North & West   
69       West Toronto                      High Park, The Junction South   
72         North York                                         Willowdale   
73    Central Toronto                                 North Toronto West   
74    Central Toronto                The Annex, North Midtown, Yorkville   
75       West Toronto                             Parkdale, Roncesvalles   
76        Mississauga              Canada Post Gateway Processing Centre   
79    Central Toronto                                         Davisville   
80   Downtown Toronto                     University of Toronto, Harbord   
81       West Toronto                                 Runnymede, Swansea   
83    Central Toronto                        Moore Park, Summerhill East   
84   Downtown Toronto          Kensington Market, Chinatown, Grange Park   
86    Central Toronto  Summerhill West, Rathnelly, South Hill, Forest...   
87   Downtown Toronto  CN Tower, King and Spadina, Railway Lands, Har...   
92   Downtown Toronto                                     Stn A PO Boxes   
94          Etobicoke                                          Northwest   
96   Downtown Toronto                        St. James Town, Cabbagetown   
97   Downtown Toronto             First Canadian Place, Underground city   
98          Etobicoke      The Kingsway, Montgomery Road, Old Mill North   
99   Downtown Toronto                               Church and Wellesley   
100      East Toronto              Business reply mail Process

Cluster 3 covers areas of greatest mixture of coffee shop, cafe, restaurant, hotels and bars. Again from the map, these are closer to the Toronto city centre. Hence the assumption is that cluster 3 is __Downtown and Business__ areas.

__Cluster 4__

In [94]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1,2] + list(range(6, toronto_merged.shape[1]))]]

Borough                                       Neighborhood  \
0    North York                                          Parkwoods   
5     Etobicoke                                   Islington Avenue   
11    Etobicoke  West Deane Park, Princess Gardens, Martin Grov...   
14    East York                                   Woodbine Heights   
21         York                                Caledonia-Fairbanks   
22  Scarborough                                             Woburn   
27   North York                                  Hillcrest Village   
50   North York                                      Humber Summit   
58  Scarborough                        Birch Cliff, Cliffside West   
88    Etobicoke       New Toronto, Mimico South, Humber Bay Shores   
93    Etobicoke                             Alderwood, Long Branch   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                   Park         Shopping Mall     Convenience Store   
5               Pharmacy     Convenience Store                Bakery   
11                  Park           Pizza Place                 Hotel   
14                  Park           Coffee Shop        Sandwich Place   
21              Pharmacy              Bus Stop                  Park   
22                  Park           Coffee Shop    Chinese Restaurant   
27           Coffee Shop                  Park              Pharmacy   
50     Electronics Store   Arts & Crafts Store              Pharmacy   
58                  Park              Gym Pool                  Café   
88                  Park                Bakery     Indian Restaurant   
93        Discount Store              Pharmacy                  Park   

    4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                Pharmacy              Bus Stop     Fish & Chips Shop   
5           Shopping Mall           Golf Course                  Park   
11             Restaurant        Clothing Store         Grocery Store   
14            Pizza Place       Thai Restaurant           Curling Ice   
21            Pizza Place    Mexican Restaurant                  Bank   
22               Pharmacy     Mobile Phone Shop     Indian Restaurant   
27                 Bakery     Korean Restaurant         Shopping Mall   
50                 Bakery         Shopping Mall                  Park   
58  General Entertainment     Convenience Store       Thai Restaurant   
88           Liquor Store          Skating Rink   Fried Chicken Joint   
93      Convenience Store           Pizza Place           Coffee Shop   

   7th Most Common Venue 8th Most Common Venue        9th Most Common Venue  \
0            Supermarket     Food & Drink Shop         Fast Food Restaurant   
5           Skating Rink         Grocery Store                         Café   
11                   Gym                  Bank            Fish & Chips Shop   
14                  Café          Skating Rink               Cosmetics Shop   
21           Coffee Shop         Grocery Store               Discount Store   
22  Fast Food Restaurant   Dumpling Restaurant  Eastern European Restaurant   
27                  Bank    Chinese Restaurant                Grocery Store   
50    Italian Restaurant                  Bank                  Pizza Place   
58        Ice Cream Shop            Restaurant              College Stadium   
88          Dessert Shop    Italian Restaurant                Grocery Store   
93                   Pub        Sandwich Place                        Trail   

   10th Most Common Venue  
0            Skating Rink  
5              Playground  
11     Mexican Restaurant  
14           Liquor Store  
21    Japanese Restaurant  
22      Electronics Store  
27             Restaurant  
50                  Field  
58                  Diner  
88               Pharmacy  
93          Shopping Mall

Cluster 4 covers areas of parks, malls, stores, pharmacy and retails store. Looking at the map the are outside of the Toronto city centre. Cluster 4 can be assumed to be __Retail Parks__ area.

__Cluster 5__

In [95]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1,2] + list(range(6, toronto_merged.shape[1]))]]

Borough              Neighborhood 1st Most Common Venue  \
45  North York  York Mills, Silver Hills                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
45                  Pool                   Zoo        Farmers Market   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
45   Dumpling Restaurant  Eastern European Restaurant     Electronics Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
45  Ethiopian Restaurant           Event Space     Falafel Restaurant

Cluster 5 covers areas of park, pool and zoo. Cluster 5 can be assumed to be __Recreational__.

__Cluster 6__

In [96]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 5, toronto_merged.columns[[1,2] + list(range(6, toronto_merged.shape[1]))]]

Borough                                       Neighborhood  \
6        Scarborough                                     Malvern, Rouge   
10        North York                                          Glencairn   
17         Etobicoke  Eringate, Bloordale Gardens, Old Burnhamthorpe...   
23         East York                                            Leaside   
26       Scarborough                                          Cedarbrae   
28        North York    Bathurst Manor, Wilson Heights, Downsview North   
29         East York                                   Thorncliffe Park   
32       Scarborough                                Scarborough Village   
39        North York                                    Bayview Village   
40        North York                                          Downsview   
44       Scarborough                    Golden Mile, Clairlea, Oakridge   
46        North York                                          Downsview   
49        North York           North Park, Maple Leaf Park, Upwood Park   
53        North York                                          Downsview   
56              York   Del Ray, Mount Dennis, Keelsdale and Silverthorn   
57        North York                                   Humberlea, Emery   
60        North York                                          Downsview   
63              York                      Runnymede, The Junction North   
65       Scarborough  Dorset Park, Wexford Heights, Scarborough Town...   
66        North York                                    York Mills West   
71       Scarborough                                  Wexford, Maryvale   
78       Scarborough                                          Agincourt   
82       Scarborough            Clarks Corners, Tam O'Shanter, Sullivan   
85       Scarborough  Milliken, Agincourt North, Steeles East, L'Amo...   
90       Scarborough                      Steeles West, L'Amoreaux West   
91  Downtown Toronto                                           Rosedale   

     1st Most Common Venue      2nd Most Common Venue   3rd Most Common Venue  \
6               Restaurant                      Trail             Coffee Shop   
10           Grocery Store       Fast Food Restaurant                     Gym   
17             Coffee Shop                  Pet Store              Beer Store   
23             Coffee Shop        Sporting Goods Shop  Furniture / Home Store   
26             Coffee Shop                       Bank             Gas Station   
28             Pizza Place                       Park             Coffee Shop   
29             Coffee Shop          Indian Restaurant           Grocery Store   
32          Ice Cream Shop          Convenience Store              Restaurant   
39     Japanese Restaurant              Grocery Store             Gas Station   
40             Coffee Shop      Vietnamese Restaurant           Grocery Store   
44            Intersection                     Bakery                Pharmacy   
46             Coffee Shop      Vietnamese Restaurant           Grocery Store   
49             Coffee Shop          Convenience Store      Dim Sum Restaurant   
53             Coffee Shop      Vietnamese Restaurant           Grocery Store   
56  Furniture / Home Store             Discount Store           Grocery Store   
57                  Bakery          Convenience Store             Golf Course   
60             Coffee Shop      Vietnamese Restaurant           Grocery Store   
63             Coffee Shop                       Park                 Brewery   
65             Coffee Shop                 Restaurant  Furniture / Home Store   
66                    Park                Coffee Shop            Tennis Court   
71            Intersection  Middle Eastern Restaurant             Pizza Place   
78      Chinese Restaurant              Shopping Mall                  Bakery   
82             Coffee Shop                Pizza Place                Pharmacy   
85      Chinese Restaurant                P

Cluster 6 covers areas of greatest restaurants, grocery stores and parks. Looking at the map, the cluster are generally outisde  the Toronto City centre. Hence, the assumption is that this cluster is mainly __General Residential__ areas